src/d3a/models/area/redis_dispatcher/market_notify_event_subscriber.py

```diff
@@ -0,0 +1,123 @@
+        return f"market/{self.market_id}/notify_event/response"
+
+    def response_callback(self, payload):
+        self.redis.resume()
+
+    def subscribe_to_event_responses(self):
+        self.redis.sub_to_response(self.event_response_channel_name(), self.response_callback)
+
+    def publish_event(self, event_type: MarketEvent, **kwargs):
+        for key in ["offer", "trade", "new_offer", "existing_offer"]:
+            if key in kwargs:
+                kwargs[key] = kwargs[key].to_JSON_string()
+        send_data = {"event_type": event_type.value, "kwargs": kwargs}
+        self.redis.publish(self.event_channel_name(), json.dumps(send_data))
+        self.redis.wait()
+
+
+class MarketNotifyEventSubscriber:
+    """
+    Used from the area class, subscribes to the market events and triggers the broadcast_events
+    method
+    """
+    def __init__(self, area, root_dispatcher):
+        self.area = area
+        self.root_dispatcher = root_dispatcher
+        self.redis = ResettableCommunicator()
+        self.subscribe_to_events()
+        self.futures = []
+        self.executors = []
+        self.active_trade = False
+
+    def publish_notify_event_response(self, market_id, event_type):
+        response_channel = f"market/{market_id}/notify_event/response"
+        response_data = json.dumps({"response": event_type.name.lower()})
+        self.redis.publish(response_channel, response_data)
+
+    def _cleanup_all_running_threads(self):
+        for future in self.futures:
+            while not future.done():
+                # Wait for future to finish
+                sleep(0.1)
+        for executor in self.executors:
+            executor.shutdown(wait=False)
+        self.futures = []
+        self.executors = []
+
+    def cycle_market_channels(self):
+        self._cleanup_all_running_threads()
+        try:
+            self.redis.stop_all_threads()
+        except Exception as e:
+            logging.debug(f"Error when stopping all threads when recycling markets: {e}")
+        self.redis = ResettableCommunicator()
+        self.subscribe_to_events()
+
+    def subscribe_to_events(self):
+        channels_callbacks_dict = {}
+        for market in self.area.all_markets:
+            channel_name = f"market/{market.id}/notify_event"
+
+            def generate_notify_callback(payload):
+                event_type, kwargs = self.parse_market_event_from_event_payload(payload)
+                kwargs["market_id"] = market.id
+
+                from concurrent.futures import ThreadPoolExecutor
+                executor = ThreadPoolExecutor(max_workers=1)
+
+                def executor_func():
+                    while self.active_trade is True and event_type == MarketEvent.OFFER:
+                        # Wait if there is an active trade and a new offer arrives.
+                        # This helps to avoid having 2 concurrent trades in one area.
+                        sleep(0.5)
```

Comment:
It seams like a long time to wait. Or did you find out that it takes at least as long as half a second?

Reply:
Timeouts are adapted to 0.01 seconds now